presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -86,441 +94,709 @@
+        if (partitionStatistics == null) {
+            return OptionalDouble.empty();
+        }
+        OptionalLong rowCount = partitionStatistics.getBasicStatistics().getRowCount();
+        if (rowCount.isPresent()) {
+            verify(rowCount.getAsLong() >= 0, "rowCount must be greater than or equal to zero");
+            return OptionalDouble.of(rowCount.getAsLong());
         }
-        return Estimate.of(totalNullsCount.getValue() / rowCount.getValue());
+        return OptionalDouble.empty();
     }
 
-    private Estimate calculateDataSize(Map<String, PartitionStatistics> statisticsSample, String columnName, Estimate rowCount)
+    @VisibleForTesting
+    static Optional<DoubleRange> calculateRangeForPartitioningKey(HiveColumnHandle column, Type type, List<HivePartition> partitions)
     {
-        if (rowCount.isUnknown()) {
-            return Estimate.unknown();
+        if (!isRangeSupported(type)) {
+            return Optional.empty();
         }
 
-        int knownPartitionCount = 0;
-        double knownRowCount = 0;
-        double knownDataSize = 0;
+        List<Double> values = partitions.stream()
+                .map(HivePartition::getKeys)
+                .map(keys -> keys.get(column))
+                .filter(value -> !value.isNull())
+                .map(NullableValue::getValue)
+                .map(value -> convertPartitionValueToDouble(type, value))
+                .collect(toImmutableList());
 
-        for (PartitionStatistics statistics : statisticsSample.values()) {
-            if (!statistics.getBasicStatistics().getRowCount().isPresent()) {
-                continue;
-            }
-            double partitionRowCount = statistics.getBasicStatistics().getRowCount().getAsLong();
+        if (values.isEmpty()) {
+            return Optional.empty();
+        }
 
-            HiveColumnStatistics partitionColumnStatistics = statistics.getColumnStatistics().get(columnName);
-            if (partitionColumnStatistics == null || !partitionColumnStatistics.getTotalSizeInBytes().isPresent()) {
-                continue;
-            }
+        double min = values.get(0);
+        double max = values.get(0);
 
-            knownPartitionCount++;
-            knownRowCount += partitionRowCount;
-            // Note: average column length from Hive might not translate directly into internal data size
-            knownDataSize += partitionColumnStatistics.getTotalSizeInBytes().getAsLong();
+        for (Double value : values) {
+            if (value > max) {
+                max = value;
+            }
+            if (value < min) {
+                min = value;
+            }
         }
 
-        if (knownPartitionCount == 0) {
-            return Estimate.unknown();
+        return Optional.of(new DoubleRange(min, max));
+    }
+
+    @VisibleForTesting
+    static double convertPartitionValueToDouble(Type type, Object value)
```

Comment:
This logic is repeated in multiple places. Any chance it can be moved into a helper class and reused?

Reply:
There are 3 similar places. One in `presto-hive`, one in `presto-tpch` and one in `presto-tpcds`. The code in `presto-hive` converts presto value to Double. Ones in `presto-tpch` and  `presto-tpcds` convert some statistics specific value (that is also represented as untyped Object) to double.

The difference is for example for the `INTEGER` type. In Presto `Integer` value is represented as `Long`, when the statistic value in `presto-tpch` and `presto-tpcds` for the Integer type is represented as `Integer`.